File: object-oriented-programming-projects/project-6/project/iterator.cpp


namespace SpaceCollection
{
template<class E, class C>
bool Iterator<E,C>::hasNext(){
	return (size - index > 0);
}

template<class E, class C>
E Iterator<E,C>::next(){
	if(!hasNext())
		throw MyErrors(-1);
	auto it = iterator;
	auto elem = *iterator;
	++index;
	++it;
	iterator = it;
	return elem;
}

template<class E, class C>
void Iterator<E,C>::remove(){
	if (size == 0)
		throw MyErrors(-2);
	if (typeid(vector<E> *) == typeid(conPtr)){
		--iterator;
		--index;
		--size;
		conPtr->erase(iterator);
	}
	else{
		auto it = iterator;
		--index;
		--size;
		--it;
		conPtr->erase(it);
	}
}
//	end of SpaceCollection namespace
}

File: object-oriented-programming-projects/project-8/project/Iterator.java
/**
 * Iterator class is a generic class that iterates any of the collection objects
 * depending on their container type
 * 
 * @author Tayfun
 *
 * @param <T> type that the assigned collection object holds and iterator will amnipulate
 */
public class Iterator<T>{
	private Collection<T> col;
	private int index;
	
	Iterator(){
		col = null;
		index = 0;
	}
	
	Iterator(Collection<T> collection, int in){
		col = collection;
		index = in;
	}
	/**
	 * hasNext checks if there is any more element in the container
	 * 
	 * @return the result
	 */
	boolean hasNext(){
		return (col.getSize() - index > 0);
	}
	/**
	 * next gives the next element from the container
	 * 
	 * @return the element which type is parametrized
	 */
	public T next() throws MyErrors{
		if(!hasNext())
			throw new MyErrors(-1);
		T elem = col.getElem(index);
		++index;
		return elem;
	}
	/**
	 * remove erases the last reached element
	 * 
	 */
	void remove() throws MyErrors{
		if (col.getSize() == 0)
			throw new MyErrors(-2);
		T[] constcon = col.getContainer();
		T[] tempcon = (T[]) new Object[col.getSize()-1];
		--index;
		for(int i=0; i<index; ++i){
			tempcon[i] = constcon[i];
		}
		for(int i=index; i<col.getSize()-1; ++i){
			tempcon[i] = constcon[i+1];
		}
		col.setContainer(tempcon);
		//++index;
		col.setSize(col.getSize()-1);
	}
}
